go/pkg/cas/upload.go

```diff
@@ -39,6 +65,361 @@ type DigestStat struct {
 
 // Upload uploads all inputs. It exits when inputC is closed or ctx is canceled.
 func (c *Client) Upload(ctx context.Context, inputC <-chan *UploadInput) (stats *TransferStats, err error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	// Do not exit until all sub-goroutines exit, to prevent goroutine leaks.
+	defer eg.Wait()
+
+	u := &uploader{
+		Client: c,
+		eg:     eg,
+
+		fsSem: semaphore.NewWeighted(int64(c.FSConcurrency)),
+	}
+
+	// Start processing input.
+	eg.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case in, ok := <-inputC:
+				if !ok {
+					return nil
+				}
+				u.eg.Go(func() error {
```

Comment:
Is it fine to use eg here?

```suggestion
				eg.Go(func() error {
```

Reply:
done